### PR TITLE
agent: Handle correctly state when CEP is present in multiple CESs

### PR DESCRIPTION
**Pull request by alan-kut**
There are condition possible in which CEP changes CES. This leads to CEP being present in multiple CESs for some time. In such cases the standard logic may not work as it always expect to have a single CEP representation.

This commit changes to logic to handle multiple CEPs properly.

```release-note
Fix incorrect network policy ebpf setup that may lead to incorrect packets denies when CEP is present in multiple CES
```

### CODE REVIEW

**Review by pchaigno — COMMENTED** [review]
> agent: Fix a when CEP is present in multiple CESs, e.g. when it's transferred between CESs.

What's the user-visible impact of this bug?

**Comment by alan-kut** [comment]
> > agent: Fix a when CEP is present in multiple CESs, e.g. when it's transferred between CESs.
> 
> What's the user-visible impact of this bug?

CEP missing from the ipcache.

Current state:
There is a CEP being moved from CES1 to CES2 and agent received the following events in the following order:
- Add CES2 with CEP -> endpointUpdated(nil, CEP from CES2)
- Add CES1 with CEP -> endpointUpdated(nil, CEP from CES1)
- Update CES1 remove CEP -> endpointDeleted(CEP from CES1)

The end state is that CEP is missing from the ipcache.
In the middle CEP is not updated properly as all the calls assume there is just one CEP and it's not updated.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: You can use `CEPName` here instead of `ces.Namespace + "/" + epName`, right?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: For me it would be easier to read if this variable is named `cep` instead `c`.
Below it's harder to follow to understand what `c` is, especially because there are also `cs`, `ces` and `ep`

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Same as in `OnAdd()`, I'd prefer it to be named `cep` instead of `c`.
If you agree, please also change it in other places below.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: `if !needUpdate { return }`

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:157]: Should the `cesMutex` be used when reading or writing to `currentCES`?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Use `newCEP` and `oldCEP` instead of `newC` and `oldC` for readability and consistency.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: This scales by the number of CEPs (total pods in the cluster), right?
The map has the same number of entries as before, but now it can map to potentially two CESs, that's all?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Typo: "ot"

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: When is `cesMutex` used?
Can you add a comment here about it?

**Review by dlapcevic — APPROVED** [review]

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: comment is obsolete now?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: extra forward slash `/`

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (skmatti) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:157]: +1, probably create a separate method this?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: assertUpdate or equalUpdate?

same for other similar functions

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Fatal here so that test exits here? If the test fails, it most likely clobber followup test steps too.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Fatal here so that test exits here?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Fatal here so that test exits here?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Fatal here so that test exits here?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Add?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: add one more step here - add for old CES?

**Review by skmatti — APPROVED** [review]

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Yes

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Yes

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:157]: Good catch, thanks. Moved locking to higher methods.

Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:157]: I covered also 
cs.epWatcher.endpointDeleted and cs.epWatcher.endpointUpdated with the mutex.

In case of race condition it would be possible that two updates to the same CEP are processed in one order but epWatcher calls are called in the other.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Fixed

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Fixed

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Done, PTAL

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Yes

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Fixed

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go]: Done

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: I don't see the changes. Perhaps you didn't push them?

Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: pushed now

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:157]: lgtm

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (dlapcevic) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: lgtm

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (skmatti) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: getCEPLocked?

**Review by skmatti — COMMENTED** [review]
lgtm

**Comment by tommyp1ckles** [comment]
@alan-kut Thanks for the contribution! Is there any issues open for this bug?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: There's a bit of a convention to name internal functions that don't hold locks as "xxxLocked". i.e. you need to hold the lock to use it.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:184]: no checking if exists, could this be nil?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [inline pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Just a suggestion - I found this tricky to keep in my head while reading the code - could we make it clearer by doing something like

```go
type cesToCEPRef map[string]*types.CiliumEndpoint
// ... 
type cepToCESmap struct {
        // ... 
	cepMap map[string]cesToCEPRef
}
```

Comment (tommyp1ckles) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:195]: Also, would this be duplicating the cesName -> cep mapping data for each CEP?

would it be better to just have:
```go
cepToCES map[string]string
// single source for cesName -> CEP
cesCEPRefs map[string]*types.CiliumEndpoint
```

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Also, this could just be a regular lock.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Question, will all `*types.CiliumEndpoint`entries within a `map[string]*types.CiliumEndpoint` always point to the same *CiliumEndpoint?

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Changed to Mutex

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Fixed

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go:184]: Yes and it's WAI - if it's nil the endpointsUpdated needs to be called with nil when there is no previous endpoint. Added comment.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (alan-kut) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: Added type cesToCEPRef map[string]*types.CiliumEndpoint and added more comments to explain why we need such complicated data structure:

```
	// Maps CEP by name to a map of CES and pointer to CiliumEndpoint.
	// In rare case when CEP exists in multiple CESs it would contain all the
	// occurrences. This is needed to retrieve currently used Cilium Endpoint
	// (cepMap[cepName][currentCES[cepName]]) when update comes and to pick other
	// representation when the current one is deleted and other exist.
	// The Cilium Endpoint pointers will point to different objects from different
	// CES. They may or may not be equal to each other.
```

**Review by pchaigno — DISMISSED** [review]
The release note in the PR description needs to describe the user-visible impact (e.g., pods can't start, packets are dropped due to policy denies, etc.) such that it is useful to users when displayed in our release notes.

**Comment by alan-kut** [comment]
I updated the code with the [tommyp1ckles](https://github.com/tommyp1ckles) suggestions.

I updated the release notes - the user visible impact is that some traffic may be incorrectly denied or allowed because eBPF maps will not be set properly because ipcache inside agent is misconfigured on some cases when CEP changes CES.

I'm not aware of any issue created for this problem.

File: pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go
Comment (tommyp1ckles) [reply pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go]: thanks

**Comment by tommyp1ckles** [comment]
@alan-kut The changes look good to me. I am curious, what are the circumstances that a CEP can exist in two CES?

**Comment by alan-kut** [comment]
One example is when CEP is moved to a different CES (especially easy to trigger with identity batching when identity changes): add CEP to 2nd CES and then remove CEP from the 1st CES.

**Comment by alan-kut** [comment]
So it's not theoretical bug - I can reproduce it - but in general - it's valid, temporarily state that CEP exists in multiple CESs. Even if we remove all current occurrences it may happen in the future, as intended or due to bug.

I think long term we need proper reconciliation logic but for now this should work.

Are we good to merge?

**Comment by joestringer** [comment]
> Are we good to merge?

The requirements for merging a PR are listed here: https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#getting-a-pull-request-merged, mainly CI and review must go through.

On the CI side, Full CI has not been run yet. This needs to be done by a member of the Cilium organization. As a regular contributor, you're eligible to [become an organization member](https://github.com/cilium/community/blob/main/CONTRIBUTOR-LADDER.md#organization-member). We're still working out the exact process with this, but if you're willing to uphold the responsibilities of an organization member then I would be happy to help set you up so that you don't require another member to trigger CI for you. (In this case I'll do it, just to provide initial feedback)

On the review side, sig-k8s is required for review and it looks like @tommyp1ckles was assigned it. Thanks also to Satish / Dorde for the additional review.

**Comment by joestringer** [comment]
> some traffic may be incorrectly denied or allowed

Paul had raised that this sounds like a security issue, though I recognize also that [CiliumEndpointSlice is marked as beta in current releases](https://docs.cilium.io/en/v1.13/network/kubernetes/ciliumendpointslice_beta/#gsg-ces), so arguably it is not quite so critical to handle this with the standard security procedure. That said, we'll also discuss this amongst the @cilium/security team on how to best advise responsible disclosure in cases like this in future. Just to make you aware, if you are concerned about a security issue and want to privately disclose an issue (and even collaborate on a fix privately), you are welcome to submit an advisory via [the security tab at the top of the page](https://github.com/cilium/cilium/security).

**Review by tommyp1ckles — APPROVED** [review]

**Comment by tommyp1ckles** [comment]
@joestringer 

> was assigned it so we need that explicit approval.

Changes look good to me, just had some final questions.

If this has potential security implications, should we be back-porting this to previous releases?

**Comment by michi-covalent** [comment]
added dont-merge/waiting-for-user-feedback to wait for joe to reply to https://github.com/cilium/cilium/pull/24838#issuecomment-1518415999

**Comment by alan-kut** [comment]
I'm not the best person to asses what to backport and what not to backport.
The code didn't change much between 1.12 and main so the backport should be straightforward and the issue has security implications so I support backporting but I'll leave the decision to you.

**Comment by joestringer** [comment]
I've added a discussion item for the community meeting tomorrow. In the mean time I don't think we need to block this, we can merge and always nominate for backport later.

**Comment by joestringer** [comment]
We had some discussion during the community meeting today ([notes](https://docs.google.com/document/d/1Y_4chDk4rznD6UgXPlPvn3Dc7l-ZutGajUv1eF0VDwQ/edit#)), for now we will apply the regular policy for this PR. We'll follow up on clarifying the backporting policy for beta features going forward.

**Comment by joestringer** [comment]
Marking for backport to v1.12, v1.13 unless anyone has concerns about the risk for backporting these changes to older branches.

**Comment by ferozsalam** [comment]
Hi @alan-kut – did you assess the applicability of this issue against v1.11 in addition to v1.12? We would want to backport it to that version too unless you have reason to believe it doesn't apply.

**Comment by alan-kut** [comment]
Hi @ferozsalam - it applies and the backport is not complex. I backported it manually in my internal repo.

**Comment by julianwiedmann** [comment]
Backports for v1.12 and v1.13 are done.

**Pull request merged**
